lib/cli/src/commands/publish.rs

```
@@ -0,0 +1,647 @@
+        let manifest = std::fs::read_to_string(&manifest_path_buf)
+            .map_err(|e| anyhow::anyhow!("could not find manifest: {e}"))
+            .with_context(|| anyhow::anyhow!("{}", manifest_path_buf.display()))?;
+        let mut manifest = wapm_toml::Manifest::parse(&manifest)?;
+        manifest.base_directory_path = cwd.clone();
+
+        // See if the user is logged in and has authorization to publish the package
+        // under the correct namespace before trying to upload.
+        let (registry, username) =
+            wasmer_registry::whoami(self.registry.as_deref(), self.token.as_deref()).with_context(
+                || {
+                    anyhow::anyhow!(
+                        "could not find username / registry for registry = {:?}, token = {}",
+                        self.registry,
+                        self.token.as_deref().unwrap_or_default()
+                    )
+                },
+            )?;
+
+        let registry_present =
+            wasmer_registry::test_if_registry_present(&registry).unwrap_or(false);
+        if !registry_present {
+            return Err(anyhow::anyhow!(
+                "registry {} is currently unavailable",
+                registry
+            ));
+        }
+
+        validate::validate_directory(&manifest, &registry, cwd.clone())?;
+
+        builder.append_path_with_name(&manifest_path_buf, "wapm.toml")?;
+
+        let package = &manifest.package;
+        let modules = manifest.module.as_ref().ok_or(PublishError::NoModule)?;
+        let manifest_string = toml::to_string(&manifest)?;
+
+        let readme = package.readme.as_ref().and_then(|readme_path| {
+            let normalized_path = normalize_path(&manifest.base_directory_path, readme_path);
+            if builder
+                .append_path_with_name(&normalized_path, readme_path)
+                .is_err()
+            {
+                // TODO: Maybe do something here
```

Should we do a `log::warn!()` here so the error doesn't get silently dropped?

---

ok done